csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -15,442 +16,279 @@ namespace nvfuser::preseg_passes {
-  std::vector<IterDomain*> mapped_alloc_dom =
-      propagateAllocationDomain(in, out);
-
-  // step 2: push each mapped iterdomain
-  std::copy(
-      mapped_alloc_dom.begin(),
-      mapped_alloc_dom.end(),
-      std::back_inserter(alloc_domain));
-
-  // step 3: compute permutation
-  std::optional<AllocationOrder> permutation =
-      ir_utils::computePermutation(out->getMaybeRFactorDomain(), alloc_domain);
-
-  NVF_ERROR(
-      permutation.has_value(),
-      "allocation order propagation on broadcast op failed to compute valid permutation");
-  alloc_order_map_[out] = permutation.value();
-}
-
-void AllocationOrderInferencer::handle(BinaryOp* op) {
-  auto* out = dynamic_cast<TensorView*>(op->out());
-  if (out == nullptr) {
-    return;
-  }
-  propagateAllocationOrder(resolveAllocationOrder(op->inputs()), out);
-}
-
-void AllocationOrderInferencer::handle(TernaryOp* op) {
-  auto* out = dynamic_cast<TensorView*>(op->out());
-  if (out == nullptr) {
-    return;
-  }
-  propagateAllocationOrder(resolveAllocationOrder(op->inputs()), out);
-}
-
-void AllocationOrderInferencer::handle(PadOp* op) {
-  auto* out = dynamic_cast<TensorView*>(op->out());
-  auto* in = dynamic_cast<TensorView*>(op->in());
-  // Note: `out` from pad has rfactor domain that cannot be mapped back to
-  // `in`'s root domain. Hence we use `out`'s root domain to match permutation.
-  propagateAllocationOrder(in, out, out->getRootDomain());
-}
+  // propagate new allocation domain on dsts
+  for (TensorView* dst : dsts) {
+    // safe check when allocation domain on the entry cannot be safely mutated.
+    if (dst == nullptr || dst->hasAllocation() ||
+        fusion->getOutputAlias(dst).type != AllocationType::New) {
+      continue;
+    }
 
-void AllocationOrderInferencer::handle(ReductionOp* op) {
-  auto* out = dynamic_cast<TensorView*>(op->out());
-  auto* in = dynamic_cast<TensorView*>(op->in());
-  propagateAllocationOrder(in, out);
-}
+    // skip entry with self mapping.
+    if (hasSelfMapping(dst, exact_graph).has_value()) {
+      continue;
+    }
 
-} // namespace
+    // find a ref among srcs to be propagated to given dst
+    TensorView* ref = nullptr;
 
-// Note [ Allocation Order Propagation ]
-//
-// The propagation tries to propagate allocation order from inputs to the entire
-// fusion:
-//   1. Iterates through all inputs, looking for TensorView with allocation
-//   domain that's a permutation of its corresponding rfactor domain and record
-//   it as the allocation order of the tensor;
-//   2. Traverse the fusion IR, propagate allocation order and record results in
-//   alloc_order_map.
-std::unordered_map<const TensorView*, AllocationOrder> inferenceAllocationOrder(
-    Fusion* fusion) {
-  std::unordered_map<const TensorView*, AllocationOrder> alloc_order_map;
+    // high water mark for candidate of ref.
+    size_t non_bc_high_water_mark = 0;
+    for (const auto& iter : loop_iter_count) {
```

Comment:
For determinism, can you change this to `for each source` and `loop_iter_count.at`? As is, I suspect ties may lead to different reference tensors being picked, making debugging hard.

Reply:
> As is, I suspect ties may lead to different reference tensors being picked, making debugging hard.

I think you see it below that we requires all `tie` to have identical mapping, hence the order of how we run into ties doesn't matter.

Reply:
> we requires all tie to have identical mapping

Got it but that's subject to change, right? If/when that happens, I'm afraid nobody will remember to change this loop to be deterministic :) 

Also, given loop_iter_count is only used once, you may want to countLoopIterDomains ionside the loop. 

```
for (TensorView* src : srcs) {
  if (src has self mapping) {
    continue;
  }
  int64_t loop_iter_count = countLoopIterDomains(...);
  ...
}
```

Side note: I don't think it'll be useful here but structured binding can avoid having to write `iter.first` and `iter.second`, which are not named or typed well. E.g. 

```
for (const auto& [src,  count] : loop_iter_count) {
```